File: lzy-zuoye/PCG.m
function [x, rs, ps, alphas] = PCG(A, b, max_iter, tol, x0)

n = length(b);  % 获取问题的维度

% 初始化输出量
x = zeros(n, max_iter+1);  % 解向量在每一步的值
rs = zeros(n, max_iter+1);  % 残差向量在每一步的值
ps = zeros(n, max_iter+1);  % 搜索方向在每一步的值
alphas = zeros(1, max_iter+1);  % 步长在每一步的值

x(:,1) = x0;  % 设置初始解
r = b - A * x(:,1);  % 计算初始残差
p = r;  % 设置初始搜索方向

rs(:,1) = r;  % 存储初始残差
ps(:,1) = p;  % 存储初始搜索方向

for i = 1:max_iter
    alpha = (r' * r) / (p' * A * p);  % 计算步长
    x(:,i+1) = x(:,i) + alpha * p;  % 更新解
    r_new = r - alpha * A * p;  % 计算新残差
    
    if norm(r_new) < tol  % 检查收敛性
        break;
    end
    
    beta = (r_new' * r_new) / (r' * r);  % 计算beta
    p = r_new + beta * p;  % 更新搜索方向
    
    r = r_new;  % 更新残差
    
    rs(:,i+1) = r;  % 存储新残差
    ps(:,i+1) = p;  % 存储新搜索方向
    alphas(:,i) = alpha;  % 存储步长
end

% 根据实际迭代次数截取输出量
x = x(:, 1:i+1);
rs = rs(:, 1:i+1);
ps = ps(:, 1:i+1);
alphas = alphas(:, 1:i);
end

File: 第一次作业/B230200148-韩林慧/PCG.py
import numpy as np

def diagonal_matrix(n1, n2, n3, n):
    """
    矩阵生成
    :param n1:对角线下
    :param n2:对角线
    :param n3:对角线上
    :param n:维度
    :return:生成对称矩阵
    """
    array_a = np.diag([n1] * (n - 1), -1)
    l = range(n2+1)
    l = l[1::]
    array_b = np.diag(l)
    array_c = np.diag([n3] * (n - 1), 1)
    matrix_A = array_a + array_b + array_c
    return matrix_A

def PCG(x,A,b,Max,error,P):
    r = b - np.dot(A,x) # x:初值, xA = b
    z = np.dot(np.linalg.inv(P),r) # P:预处理矩阵，对称正定
    beta = np.linalg.norm(z,2)
    if beta < error:
        print(x)
    else:
        for m in range(Max):
            c = np.dot(r.T, z)
            if m == 0:
                p = z
                cc = c
            elif m > 0:
                mju = c / cc
                p = z + np.dot(mju,p)
            q = np.dot(A,p)
            Xi = c / np.dot(p.T,q)
            x = x + np.dot(Xi,p)
            r = r - np.dot(Xi,q)
            relres = np.linalg.norm(r,2) / beta
            print(relres)
            if relres < error:
                break
            else:
                z = np.dot(np.linalg.inv(P),r)
                cc = c
        if relres < error:
            return x
        else:
            return 0


A = diagonal_matrix( -1,100,-1,100)
# A = np.random.rand(100,100)
b = np.dot(A,np.ones(100))
Max = 30
error = 1e-8
x = np.zeros(100, dtype = float)
a = range(1,101)
P = np.diag(a)
y = PCG(x,A,b,Max,error,P)
print(A)
print(b)
print(y)